rfcs/0003-gene-sets/api.md

```diff
@@ -0,0 +1,400 @@
+
+The name of the exported file will be based on the user name and a timestamp rounded to the nearest second.
+
+### Keeping Consensus
+
+This feature will require processing multiple updates to the same gene set, due to our hosting architecture this creates
+the potential for race conditions and concerns about keeping the data in sync between the database, server and client.  
+To ensure data consistency, when the user is interacting with the data, the client is always the source of truth.
+If what is stored in the database doesnt match what the user is seeing, the data in the database will be discarded and
+recreated based on the state of the client.
+
+The server will ensure the data is kept in sync through a consensus token or counter. Everytime a gene set is created
+in the database it will have a consensus_counter field set to 0000. This will be sent back to the client along with the
+gene set name and uuid. When the client updates a gene set (via a PUT request) it will send this counter back to the
+server as part of the request body. The server will confirm that the counter sent from the client matches the counter
+stored in the database for that gene set.
+
+- If the values match, the server will update the gene set as specified in the PUT
+  request and update the counter by incrementing it by 1. The response to the client will include the updated counter value
+  to be used in future PUT requests for that gene set.
+- If the counter stored in the database does not match the counter sent by the client the server will immediately send
+  an error message to the client, specifying the name and uuid of the gene set that is out of sync. The client will
+  then send over the entire gene set including all linked genes as a POST request. Because the gene set name is already
+  taken for that user/dataset, the server will delete the currently stored gene set and any associated genes and recreate
+  it based on the information included in the POST request. This will generate a new UUID for the gene set, and reset the
+  consensus counter.
+  In the case that the server times out or does not respond to a PUT request from the client, the client can retry or
+  just send a POST request to recreate the gene set.
+  Because recreating the gene set assigns it a new UUID, any pending or high latency PUT requests sent prior to the update
+  be ignored because they reference a UUID that no longer exists.
+
+In the desktop version of the application the gene sets (and genes) will be stored locally as a csv under the name of
+the dataset and the user session. On updates the application will create and store a new csv, maintaining the last
+10 updates (this will match the current functionality for annotations in the desktop environment)
+
+### Data Lifecycle
+
+#### Creation
+
+Gene sets are created via a POST request to `/dataset_name/api/v0.1/gene_sets/`. This request is sent
```

Comment:
gene_sets of var_sets.   We had a discussion about this at the last design meeting, not sure if we landed on a a decision.

Reply:
If we update the api to be var_sets do you think we should also update the table in the db?

Reply:
Yes, that make sense.